pkg/config/handlerConfigurer_test.go

```diff
@@ -0,0 +1,372 @@
+		if tt.tmplCnfgrMtdShouldPanic && !tt.handlers["hndlr"].isBroken {
+			t.Error("The handler should be marked as broken.")
+		}
+
+		if tt.wantErr == "" {
+			if err != nil {
+				t.Errorf("got err %v\nwant <nil>", err)
+			}
+			ensureInterfacesInConfigCallsValid(t, tt.expectCallTrackInfo, actualCallTrackInfo)
+		} else {
+			if err == nil || !strings.Contains(err.Error(), tt.wantErr) {
+				t.Errorf("got error %v\nwant %v", err, tt.wantErr)
+			}
+		}
+	}
+}
+
+func TestInferTypes(t *testing.T) {
+	tests := []struct {
+		name         string
+		constructors map[string]*pb.Constructor
+		tmplRepo     tmpl.Repository
+		want         map[string]proto.Message
+		wantError    string
+	}{
+		{
+			name:         "SingleCnstr",
+			constructors: map[string]*pb.Constructor{"inst1": {"inst1", "tpml1", &empty.Empty{}}},
+			tmplRepo:     newFakeTmplRepo(nil, &wrappers.Int32Value{Value: 1}, true),
+			want:         map[string]proto.Message{"inst1": &wrappers.Int32Value{Value: 1}},
+			wantError:    "",
+		},
+		{
+			name: "MultipleCnstr",
+			constructors: map[string]*pb.Constructor{"inst1": {"inst1", "tpml1", &empty.Empty{}},
+				"inst2": {"inst2", "tpml1", &empty.Empty{}}},
+			tmplRepo:  newFakeTmplRepo(nil, &wrappers.Int32Value{Value: 1}, true),
+			want:      map[string]proto.Message{"inst1": &wrappers.Int32Value{Value: 1}, "inst2": &wrappers.Int32Value{Value: 1}},
+			wantError: "",
+		},
+		{
+			name:         "ErrorDuringTypeInfr",
+			constructors: map[string]*pb.Constructor{"inst1": {"inst1", "tpml1", &empty.Empty{}}},
+			tmplRepo:     newFakeTmplRepo(fmt.Errorf("error during type infer"), nil, true),
+			want:         nil,
+			wantError:    "cannot infer type information",
+		},
+	}
+	ex, _ := expr.NewCEXLEvaluator(expr.DefaultCacheSize)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			hc := handlerConfigurer{typeChecker: ex, tmplRepo: tt.tmplRepo}
+			v, err := hc.inferTypes(tt.constructors)
+			if tt.wantError == "" {
+				if err != nil {
+					t.Errorf("got err %v\nwant <nil>", err)
+				}
+				if !reflect.DeepEqual(tt.want, v) {
+					t.Errorf("got %v\nwant %v", v, tt.want)
+				}
+			} else {
+				if err == nil || !strings.Contains(err.Error(), tt.wantError) {
+					t.Errorf("got error %v\nwant %v", err, tt.wantError)
+				}
+			}
+
+		})
+	}
+}
+
+func TestDedupeAndGroupInstances(t *testing.T) {
+	tests := []struct {
+		name         string
+		actions      []*pb.Action
+		constructors map[string]*pb.Constructor
+		handlers     map[string]*HandlerBuilderInfo
+		want         map[string]instancesByTemplate
+		wantError    string
+	}{
+		{
+			name:         "SimpleNoDedupeNeeded",
+			handlers:     map[string]*HandlerBuilderInfo{"hndlr1": nil},
+			constructors: map[string]*pb.Constructor{"i1": {"i1", "tpml1", nil}},
+			actions:      []*pb.Action{{"hndlr1", []string{"i1"}}},
+			want: map[string]instancesByTemplate{
+				"hndlr1": {map[string][]string{"tpml1": {"i1"}}},
+			},
+			wantError: "",
```

Comment:
no need to have `wantError: ""` in these cases. that's the zero value for string anyway.